reference/Tensor.cpp

```diff
@@ -0,0 +1,365 @@
+  if (elementType.isUnsignedInteger(4) || elementType.isUnsignedInteger(8)) {
+    auto elementData = reinterpret_cast<uint8_t *>(elementPtr);
+    auto value = getIntegerValue(element);
+    *elementData = (uint8_t)value.getZExtValue();
+    return;
+  }
+
+  if (elementType.isUnsignedInteger(16)) {
+    auto elementData = reinterpret_cast<uint16_t *>(elementPtr);
+    auto value = getIntegerValue(element);
+    *elementData = (uint16_t)value.getZExtValue();
+    return;
+  }
+
+  if (elementType.isUnsignedInteger(32)) {
+    auto elementData = reinterpret_cast<uint32_t *>(elementPtr);
+    auto value = getIntegerValue(element);
+    *elementData = (uint32_t)value.getZExtValue();
+    return;
+  }
+
+  if (elementType.isUnsignedInteger(64)) {
+    auto elementData = reinterpret_cast<uint64_t *>(elementPtr);
+    auto value = getIntegerValue(element);
+    *elementData = (uint64_t)value.getZExtValue();
+    return;
+  }
+
+  // Handle complex types.
+  if (elementType.isa<ComplexType>()) {
+    auto complexElemTy = elementType.cast<ComplexType>().getElementType();
+
+    if (complexElemTy.isF32()) {
+      auto elementData = reinterpret_cast<std::complex<float> *>(elementPtr);
+      auto complexValue = getComplexValue(element);
+      *elementData = std::complex<float>(complexValue.real().convertToFloat(),
+                                         complexValue.imag().convertToFloat());
+      return;
+    }
+
+    if (complexElemTy.isF64()) {
+      auto elementData = reinterpret_cast<std::complex<double> *>(elementPtr);
+      auto complexValue = getComplexValue(element);
+      *elementData =
+          std::complex<double>(complexValue.real().convertToDouble(),
+                               complexValue.imag().convertToDouble());
+      return;
+    }
+  }
+
+  llvm_unreachable("Unsupported element type");
+}
+
+void Tensor::print(raw_ostream &os) const {
+  getType().print(os);
+  os << " {\n";
+  for (auto i = 0; i < getNumElements(); ++i) {
+    os << "  " << get(i) << "\n";
+  }
+  os << "}";
+}
+
+void Tensor::dump() const {
+  print(llvm::errs());
+  llvm::errs() << "\n";
+}
+
+Tensor makeTensor(ShapedType type, ArrayRef<StringRef> strData) {
+  auto elemType = type.getElementType();
+
+  if (auto complexTy = elemType.dyn_cast<ComplexType>()) {
+    auto complexElemTy = complexTy.getElementType();
+    auto floatType = complexElemTy.dyn_cast<FloatType>();
+    if (!floatType)
+      llvm_unreachable("Unsupported element type for commplex type");
+
+    auto floatValues = llvm::to_vector(
+        llvm::map_range(strData, [&](StringRef strNum) -> APFloat {
+          return APFloat(floatType.getFloatSemantics(), strNum);
+        }));
+
+    auto complexData = llvm::makeArrayRef(
+        reinterpret_cast<std::complex<APFloat> *>(floatValues.data()),
+        floatValues.size() / 2);
+    return Tensor(DenseElementsAttr::get(type, complexData));
+  }
+
+  if (auto floatType = elemType.dyn_cast<FloatType>()) {
```

Comment:
I think we had a useful comment here internally:

```
  // Unlike parsing Int literals, we are not using parseAttribute for parsing
  // Float literals mainly because it does not seem to parse special float
  // values like nan, +/-inf.
```